PROJECT_PROPOSAL.md

```diff
@@ -0,0 +1,99 @@
+
+```lean
+theorem left_distrib {R : Type v} [Mul R] [Add R] [LeftDistribClass R] (a b c : R) :
+a * (b + c) = a * b + a * c
+```
+
+And figure out the correct way to invoke this in the local context, correctly instantiating all the generic arguments, keeping track of customizable operator precedence and associativity rules, etc etc. If they apply it incorrectly, they can expect a long and hard-to-understand error message.
+
+We propose to, instead, build direct-manipulation, [**what you see is what you get**](https://en.wikipedia.org/wiki/WYSIWYG) interfaces for various kinds of algebra. These should provide users with large, readable visualizations of proof state -- traditional terms, commutative diagrams, etc. 
+
+Very simple interactions should be possible by *direct manipulation* of these visualizations. For example, in ring theory, we can design interactions for each of the ring axioms:
+
+- `a + b = b + a`
+- `a * (b + c) = a * b + a * c`
+- `a + 0 = a`
+- ...
+
+An example sequence of simple interactions might look like this:
+
+![example interaction](./images/dragging.jpg)
+![example interaction 2](./images/dragging2.jpg)
+
+Notice how the system provides continual feedback. When the user attempts to make a valid move, the system reassures them it is valid and records it. Invalid moves are simply not possible. The system invites experimentation, encouraging the user to freely explore.
+
+Of course, this model of direct manipulation is unlikely to work for anything but the simplest identities. The system will also need to provide affordances for modularity -- UIs for recording complex chains of rewrites, and applying them to more complicated proof states. These will necessarily be more complicated to use, but we hope that they can still be in the spirit of direct manipulation.
+
+As a sanity check, the proofs from this system should be able to be compiled to run in a traditional, trusted proof assistant. More exotically, we might compile them to other things, like visualizations. For example, we can compile proofs about low-degree polynomials in the real numbers to proofs about areas of rectangles. The left-distributivity axiom now takes the form:
+
+![representation](./images/representation.jpg)
+
+This sort of compilation is what mathematicians call a "representation" of one sort of mathematics -- in this case, algebra -- in terms of another sort of mathematics -- in this case, geometry. Much of modern algebra grew out of the study of representations. Building these kinds of representations into the system can help anchor students' understanding of abstract subject matter. 
```

Comment:
This is almost certainly too vague but I didn't want to get bogged down in technical details.

I want to think of proofs as some sort of n-morphism and models of these proofs as some sort of functor.

In the example above, the identity holds regardless of the lengths of the line segments a,b,c. To show this to the user, we could let them manually modify the lengths of the segments, and see that the identity still holds.

To be really slick, you could bake this information into the target category of the visualization. An object of the category is then, not just a diagram, but a parameterized family of diagrams. And the UI lets you directly modify the parameters. But I haven't thought very hard about this.